text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+impl LedgerSerializer for MyAsset {
+    fn to_buffer(&self) -> Vec<u8> {
+        /// my own form of serialization eg convert to XML
+    }
+
+    fn from_buffer(buffer: Vec<u8>) -> MyAsset{
+        /// my own form of serialization eg convert to XML
+    }
+}
+```
+
+By using the `from` and `into` standard patterns in then can be used. For example.
+
+```rust
+    let new_asset = MyAsset::new(my_assset_id,value);
+    world.create(State.from(new_asset));
+```
+
+## Transactional Information
+
+### Current Transaction 
+
+To obtain the current transaction and the essential information:
+
+```rust
+    let transaction = Transaction::current_transaction();
+    transaction.get_timestamp();
+    transaction.get_id();
+    transaction.get_mspid();
+```
+
+`get_mspid` is important for handling situations involving private data and being able to determine the originating organization. 
+
+### Events
+
+Currently, a single event payload and name can be produced from a chaincode. Feedback has indicated that at a minimum it should be possible to have multiple names and payloads.
+
+The transaction, therefore, can support this
+
+```rust
+    transaction.set_event(evtName, eventData);
+    transaction.get_event(evtName, eventData);
+```
+
+These will then be put into a JSON Object structure and returned. The further specification will be left to later work. 
+
+### Identity
+
+Information about the identity that requested the transaction can be obtained:
+
+```rust
+        let identity = transaction.get_client_identity();
+        identity.attribute_exists(attrName,attrValue);
+        identity.attribute_value(attrName);
+        identity.getID()
+        identity.getMSPID();
+```
+
+Will validate that this is the appropriate set of the accessors to support against real use-cases. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Wasm Runtime
+
+These are the major components within the runtime.
+
+- a Wasm runtime engine responsible for the execution of the Wasm binary
+- grpc communications with the peer
+- data marshalling for transfer of data and requests into and out of the Wasm engine
+- Along with required code to bootstrap the whole Wasm engine and load in the binary
+
+### Wasm Engine
+Several different implementations are available; initial prototypes have used the LifeVM, but for final implementation, Wasmer is being considered. The WebAssembly landscape is still developing so there is some risk with changing standards and implementations, however, some of the new standards could be of benefit in the future.
+
+### grpc Peer Communications
+This will use the go chaincode shim today to provide connectivity to the peer.
+
+### Data Marshalling
+Currently, the API to import/export functions and data between the host and the executing code in the wasm runtime is limited to primitive data types only. A data and function marshalling interface is required, waPC will be used to provide this. 
```

Comment:
need to define wapc, https://github.com/wapc

very interesting to see this solution to the procedure call issue. Amazing how fast wasm is advancing. A bit like watching language advancement in fast-forward.